### DISCUSSION TITLE
Publish storage deals via a smart contract

### DISCUSSION THREAD

**Discussion by jennijuju** [filecoin-project/community#584] [Developers]
With FEVM, filecoin storage deals can now be published via smart contracts - by calling the [PublishStorageDeals method in the filecoin storage market actor (f05)](https://github.com/longfeiWan9/BuiltIn-Actor-API/blob/main/Storage%20Market%20Actor.md#publishstoragedeals. ). However, today in the Filecoin protocol, only a storage provider(miner) can publish storage deals onchain using their control/owner address. This means a smart contract must be set as a storage provider's control address first before publish deals on the provider's behalf.

Here is a potential solution for enabling deal publishments via smart contracts which can further extend to a DataDao, proposed by @raulk , originally [here](https://github.com/filecoin-project/FIPs/discussions/401#discussioncomment-4712380):
1. A DataDAO allows miners to register to participate as storage providers in the DAO.
2. When an SP registers, the DataDAO creates a “Participating Storage Provider” (PSP) contract via CREATE2 and returns its address. This is the gateway for the miner to participate in the DAO.
3. The SP adds the PSP as a control address via [`ChangeWorkerAddress`](https://github.com/longfeiWan9/BuiltIn-Actor-API/blob/main/Miner.md#changeworkeraddress) to its Miner actor, thus allowing it to call [`PublishStorageDeals`](https://github.com/longfeiWan9/BuiltIn-Actor-API/blob/main/Storage%20Market%20Actor.md#publishstoragedeals) on its behalf.
4. When the DataDAO decides it wants to make a deal with the miner, it internally creates a DealProposal and notifies the PSP contract, which emits an event that the miner is watching (through some application).
5. When N number of deals have been made, the miner decides to commit to storing them, and sends a message to the PSP, which calls PSD on its behalf with the accumulated DealProposals (given that it’s a controlling address).
6. The miner now downloads the data through whatever mechanism the DataDAO designates (IPFS, Boost, HTTP, etc.)
7. The sealing process starts and progresses (Precommit with DealIDs, ProveCommit) until culmination (deal activation).

Would love to hear other paths to onboard storage deals to filecoin via smart contracts, and hopefully we will see a prototype here soon!

**Comment by jennijuju** [comment]
@raulk I feel like we are missing `AddFunds` in the proposed flow 🤔  i.e: an SP should add funds to the PSP, and when a call to PSD triggers, the PSP first transfers funds to f05 then publish the deals.

**Comment by kikakkz** [comment]
@jennijuju FIL-Peggy aims to custody whole miner to smart contract, including owner address and control addresses. May this be a part of it ? But before our test we cannot set smart contract address to be miner control addresses, is that possible in new FVM version ?

**Reply by zklim** [reply]
Our project PeaceFeb is also making assumption that owner address will be able to create on-chain and it can be a smart contract in future.

**Reply by kikakkz** [reply]
currently you cannot create a miner with smart contract directly but you can create one with account then set owner to smart contract, 😄, and that's what peggy is developing~

**Reply by zklim** [reply]
> currently you cannot create a miner with smart contract directly but you can create one with account then set owner to smart contract, 😄, and that's what peggy is developing~

It is? But how do I convert Ethereum address of that smart contract to Filecoin address and vice versa?

The reason why we want this is just to retain control of funds so we can restrict fund usage and also program block rewards flow.

**Reply by kikakkz** [reply]
for eth address to fil address you can use @glif/local-managed-provider in js, or just refer to lotus code for golang
